include/up-cpp/utils/Expected.h

```diff
@@ -22,87 +23,161 @@ namespace uprotocol::utils {
-
-	constexpr T value_or(T&& default_value) const&;
-	constexpr T value_or(T&& default_value) &&;
-
-	template <typename F>
-	constexpr auto and_then(F&& f) &;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&;
-	template <typename F>
-	constexpr auto and_then(F&& f) &&;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&&;
-
-	template <typename F>
-	constexpr auto or_else(F&& f) &;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&;
-	template <typename F>
-	constexpr auto or_else(F&& f) &&;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&&;
+	std::variant<E, Unexpected<U>> storage;
+
+public:
+	constexpr Expected() { storage = E(); }
+	constexpr Expected(E&& rhs) : storage(std::move(rhs)) {}
+	constexpr Expected(Unexpected<U>&& rhs) : storage(std::move(rhs)) {}
+
+	constexpr Expected(const Unexpected<U>& rhs) {
+		std::get<Unexpected<U>>(storage).storage = std::move(rhs);
+	}
+
+	template <class X = E>
+	constexpr explicit Expected(X&& rhs) {
+		std::get<E>(storage) = std::forward<X>(rhs);
+	}
+
+	template <class X = U>
+	constexpr explicit Expected(Unexpected<X>&& rhs) {
+		std::get<E>(storage) = rhs.error();
+	}
+
+	constexpr Expected(const Expected& rhs) { storage = rhs.storage; }
+
+	constexpr bool has_value() const noexcept {
+		return std::holds_alternative<E>(storage);
+	}
+
+	constexpr explicit operator bool() const noexcept {
+		return std::holds_alternative<E>(storage);
+	}
+
+	template <class X>
+	constexpr E value_or(X&& v) const& noexcept {
+		return has_value() ? std::get<E>(storage) : v;
+	}
+
+	constexpr const E& value() const& {
+		if (!has_value())
+			throw BadExpectedAccess(
+			    "Attempt to access value() when unexpected.");
+		return std::move(std::get<E>(storage));
+	}
+
+	constexpr E value() && {
+		if (!has_value())
+			throw BadExpectedAccess(
+			    "Attempt to access value() when unexpected.");
+		return std::move(std::get<E>(storage));
+	}
+
+	constexpr const U& error() const& {
+		if (has_value())
+			throw BadExpectedAccess(
+			    "Attempt to access error() when not unexpected.");
+		return std::move(std::get<Unexpected<U>>(storage).error());
+	}
+
+	constexpr U error() && {
+		if (has_value())
+			throw BadExpectedAccess(
+			    "Attempt to access error() when not unexpected.");
+		return std::move(std::get<Unexpected<U>>(storage)).error();
+	}
+
+	constexpr E& operator*() {
+		if (!has_value())
+			throw BadExpectedAccess(
+			    "Attempt to non-const dereference expected value when "
+			    "unexpected.");
+		return std::get<E>(storage);
+	}
+
+	constexpr const E& operator*() const {
+		if (!has_value())
+			throw BadExpectedAccess(
+			    "Attempt to const dereference expected value when unexpected.");
+		return std::get<E>(storage);
+	}
+
+	constexpr E* operator->() {
+		if (!has_value())
+			throw BadExpectedAccess(
+			    "Attempt to dereference expected pointer when unexpected.");
+		return &std::get<E>(storage);
+	}
```

Comment:
I'm not sure we need to allow non-const access to the value when its not being moved out. We could probably get by with just the const `operator*`:
```suggestion
	constexpr const E& operator*() const {
		if (!has_value())
			throw BadExpectedAccess(
			    "Attempt to const dereference expected value when unexpected.");
		return std::get<E>(storage);
	}
```